diskmond.8
.\"
.TH DISKMOND 8 "November 2002" "Debian/GNU Linux"
.SH NAME
diskmond \- checks disk space for "fwdata".
.SH SYNOPSIS
diskmond
.SH DESCRIPTION
.B diskmond
checks every 60 seconds to make sure that the disk utilization for the partition that holds the psad "fwdata" file is not beyond a threshold that the administrator defines.
.SH SEE ALSO
.BR psad (8),
.BR psadwatchd (8),
.BR kmsgsd (8)
.SH AUTHOR
Michael B. Rash (mbr@cipherdyne.com)

This manual page was written by Daniel Gubser <daniel.gubser@gutreu.ch> for the Debian GNU/Linux system (but may be used by others).



ChangeLog
psad-2.4.2 (10//2015):
    - Bug fix to apply the EMAIL_ALERT_DANGER_LEVEL threshold to auto-blocking
      emails (reported by itoffshore@github).
    - Bug fix to include the META.yml file for the Unix::Syslog module. This
      issue was reported by github user itvasile as issue #26.
    - Updated IPTables::ChainMgr and IPTables::Parse to 1.4 and 1.5
      respectively.

psad-2.4.1 (05/13/2015):
    - Bug fix to honor the IGNORE_PROTOCOLS configuration variable for
      non-tcp/udp/icmp protocols. This bug was reported by Paul Versloot.
      Also extended the IGNORE_PROTOCOLS feature to match on both protocol
      name and number as well regardless of what iptables reports within log
      messages. This is so the user does not have to know what iptables will
      report (which can be inconsistent, e.g. 'TCP' vs. '2' for IGMP).
    - Added two configuration variables ENABLE_WHOIS_LOOKUPS and
      ENABLE_DNS_LOOKUPS (set to 'Y' by default) to allow whois and reverse
      DNS lookups to be controlled from the command line.
    - Bug fix for an uninitialized variable in 'psad -L' mode when auto
      blocking is enabled. This bug was reported via github issue #19 by
      gihub user 'itoffshore'.

psad-2.4.0 (03/18/2015):
    - Added support for reading syslog messages from journalctl on systems
      where syslog data is tied into systemd. psad detects journalctl by
      default, but this can be disabled via the AUTO_DETECT_JOURNALCTL
      variable. When enabled, by default the command executed by psad to
      acquire syslog data is '/bin/journalctl -f -k', but both the command
      path and the command args can be altered with the FW_MSG_READ_CMD and
      FW_MSG_READ_CMD_ARGS variables respectively.
    - Added support for systems with 'firewalld' by leveraging the
      'firewall-cmd' binary. This is done via the updated IPTables::Parse and
      IPTables::ChainMgr modules.
    - Bug fix reported by Shlomit Afgin to handle the syslog time format that
      looks like '2015-03-08T02:25:11.444012+02:00 servername kernel: ..'.
      This fix has also been extended to allow psad to handle custom time
      formats that are allowed by some syslog daemons. This is controlled via
      two new config variables CUSTOM_SYSLOG_TS_RE and CUSTOM_SYSLOG_TS_RE.
    - Updated all module dependencies in the deps/ directory.
    - Bug fix reported by Brad Rubenstein to ensure exiting with a proper exit
      status in 'psad --HUP' mode. This was also extended to ensure better
      exit status returns in other modes as well such as --Status, and --USR1.
      Fixes issue #11 on github.

psad-2.2.5 (02/09/2015):
    - Added signature to detect fwknop Single Packet Authorization (SPA)
      packets that are destined to the default UDP port 62201. More
      information about fwknop can be found here:

        https://www.cipherdyne.org/fwknop/

    - (Rinck Sonnenberg): Spotted install.pl exit code bug when uninstalling
      psad. Getting this fixed will enable a proper Puppet module to be
      developed which Rinck is working on.
    - (Rinck Sonnenberg) UFW and Logstash integration capabilities for psad
      are now available:

        Logstash:  https://gist.github.com/netson/cc6ffa751aa4610c13c0
        UFW:       https://gist.github.com/netson/c45b2dc4e835761fbccc

psad-2.2.4 (01/16/2015):
    - (Steve Murphy) Added the ability to run an external script whenever an
      IP is blocked. Two new config variables were added to support this
      feature, ENABLE_EXT_BLOCK_SCRIPT_EXEC and EXTERNAL_BLOCK_SCRIPT.
    - Bug fix to not create zombie whois processes when whois lookups take too
      long to complete for whatever reason (slow network, etc.). This fixes
      issue #15 on Github. The bug was reported by "3Turtles" to the psad
      mailing list, and Dan Dickey provided valuable input.
    - Bug fix for uninitialized variable in --fw-block-ip handling reported by
      Albert Whale (fixes issue #17).
    - Bug fix in reverse DNS resolution for IP addresses that cannot be
      properly resolved.

psad-2.2.3 (03/01/2014):
    - Added compatibility with 'upstart' init daemons with assistance from Tim
      Kramer.  This change adds a new config variable 'ENABLE_PSADWATCHD' that
      can be used to disable psadwatchd when deployed with upstart since it
      has built-in process monitoring and restarting capabilities.  By default
      psadwatchd is not enabled anymore since this variable is set to "N". The
      reason for this change is that psad is extremely stable and so almost
      never needs to be restarted in practice, and process monitoring is
      better provided via other solutions (like upstart) anyway.  In addition,
      a new init script located at init-scripts/upstart/psad.conf has been
      added that is compatible with upstart - this script is meant to be copied
      to the /etc/init/ directory.
    - (Wolfgang Breyha) Bug fix to allow VLAN interfaces and interface aliases
      in IGNORE_INTERFACES.  This fixes issue #8 on github.
    - Bug fix to not modify /etc/hosts.deny permissions when removing
      tcpwrappers auto-block rules. This issue was reported as Debian bug
      #724267 (http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=724267) and
      relayed via Franck Joncourt. Closes issue #7 on github.

psad-2.2.2 (01/13/2014):
    - Added detection for Errata Security's "Masscan" port scanner that was
      used in an Internet-wide scan for port 22 on Sept. 12, 2013 (see:
      http://blog.erratasec.com/2013/09/we-scanned-internet-for-port-22.html).
      The detection strategy used by psad relies on the fact that masscan does
      not appear to set the options portion of the TCP header, and if the
      iptables LOG rules that generate log data for psad are built with the
      --log-tcp-options switch, then no options in a SYN scan can be seen.
      This is not to say that other scanning software always sets TCP options -

diskmond.pl
#!/usr/bin/perl -w
#
#########################################################################
#
# File: diskmond
#
# Purpose: diskmond checks every 60 seconds to make sure that the disk
#          utilization for the partition that holds the psad "fwdata"
#          file is not beyond a threshold that the administrator
#          defines.
#
# Author: Michael B. Rash (mbr@cipherdyne.com)
#
# Credits:  (see the CREDITS file)
#
# Version: 1.0.0-pre4
#
# Copyright (C) 1999-2002 Michael B. Rash (mbr@cipherdyne.com)
#
# License (GNU Public License):
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307
#    USA
#
#########################################################################
#
# $Id$
#

use lib '/usr/lib/psad';
use Psad;
use Sys::Hostname 'hostname';
use File::stat;
use File::Copy;
use File::Path;
use POSIX 'setsid';
use Getopt::Long 'GetOptions';
use strict;

### establish the default path to the config file (can be
### over-ridden with the -c <file> command line option.
my $CONFIG_FILE = '/etc/psad/psad.conf';

### handle command line arguments
die " @@@ Specify the path to the psad.conf file with " . 
    "\"-c <file>\".\n\n" unless (GetOptions (
    'config=s' => \$CONFIG_FILE
));

### read in the configuration file
my ($Config_href, $Cmds_href) = &Psad::buildconf($CONFIG_FILE);

### make sure the configuration is complete
&check_config();

my %Config = %$Config_href;
my %Cmds   = %$Cmds_href;
#my $Archive_files_aref = $Config{'ARCHIVE_FILES'};

### Check to make sure the commands specified in the config section
### are in the right place, and attempt to correct automatically if not.
&Psad::check_commands(\%Cmds);

&Psad::unique_pid($Config{'DISKMOND_PID_FILE'});

### install WARN and DIE handlers
$SIG{'__WARN__'} = \&Psad::warn_handler;
$SIG{'__DIE__'}  = \&Psad::die_handler;

my $pid = fork;
exit if $pid;
die " ... @@@  $0: Couldn't fork: $!" unless defined($pid);
POSIX::setsid() or die " ... @@@  $0: Can't start a new session: $!\n";

### write the pid to the pid file
&Psad::writepid($Config{'DISKMOND_PID_FILE'});

### initialize partition usage
my $usage     = 0;
my $email_ctr = 0;
my $hostname  = hostname;

my $config_mtime = stat($CONFIG_FILE)->mtime;
#====================== main =======================
### main loop
for (;;) {
    ### See if we need to import any changed config variables
    &check_import_config(\$config_mtime, $CONFIG_FILE);

    $usage = &get_usage();
    ### Check to see if we need to start archiving
    if ($usage >= $Config{'MAX_DISK_PERCENTAGE'}) {
        &rm_data();
        open DW, "> $Config{'PSAD_DIR'}/disk_limit";
        print DW "psad: disk utilization above $Config{'MAX_DISK_PERCENTAGE'}%\n";
        print DW "      on disk partition that holds $Config{'PSAD_DIR'}\n";
        print DW "      diskmond removed all ip directories in " .
            "$Config{'PSAD_DIR'}.\n";
        close DW;
        &Psad::sendmail("psad: disk utilization above $Config{'MAX_DISK_PERCENTAGE'}%",
            "$Config{'PSAD_DIR'}/disk_limit", $Config{'EMAIL_ADDRESSES'},
            $Cmds{'mail'});
        $email_ctr++;
        if ($email_ctr > 50) {
            &Psad::sendmail("psad: diskmond giving up, stopping psad.",
                $Config{'EMAIL_ADDRESSES'}, $Cmds{'mail'});
        }
    } else {
        $email_ctr = 0;
    }
    ### check disk usage every $CHECK_INTERVAL seconds
    sleep $Config{'DISKMOND_CHECK_INTERVAL'};
}
exit 0;
#===================== end main ====================

sub get_usage() {
    my @df_data = `$Cmds{'df'} $Config{'PSAD_DIR'}`;
    my ($prcnt) = ($df_data[$#df_data] =~ /(\d+)%/);
    return $prcnt;
}

sub rm_data() {
    chdir $Config{'PSAD_DIR'} or die
        " ... @@@ Could not chdir $Config{'PSAD_DIR'}: $!";

    &rm_scanlog($Config{'PSAD_DIR'});
    &rm_scanlog($Config{'SCAN_DATA_ARCHIVE_DIR'});

    if (-d $Config{'SCAN_DATA_ARCHIVE_DIR'}) {
        rmtree $Config{'SCAN_DATA_ARCHIVE_DIR'};
        mkdir $Config{'SCAN_DATA_ARCHIVE_DIR'}, 0500;
    }
    if (-e $Config{'FW_DATA'}) {
        open F, "> $Config{'FW_DATA'}";
        close F;
    }
    if (-e "$Config{'SCAN_DATA_ARCHIVE_DIR'}/fwdata_archive") {
        open F, "> $Config{'SCAN_DATA_ARCHIVE_DIR'}/fwdata_archive";
        close F;
    }
    return;
}

sub rm_scanlog() {
    my $dir = shift;
    chdir $dir or die " ... @@@ Could not chdir($dir): $!\n";
    opendir D, $dir or
        die " ... @@@ Could not open dir: $dir: $!";
    my @files = readdir D;
    closedir D;
    shift @files; shift @files;
    for my $file (@files) {
        if ($file =~ /\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}/ && -d $file) {
            ### we found a directory like /var/log/psad/<ip>, which
            ### contains the scanlog file for this ip
            if (-e "${file}/scanlog") {
                open F, "> ${file}/scanlog";
                close F;
            }
        }
    }
    return;
}

sub check_import_config() {
    my ($mtime_ref, $file) = @_;
    my $mtime_tmp = stat($file)->mtime;
    if ($mtime_tmp != $$mtime_ref) {  ### the file was modified, so import

        ($Config_href, $Cmds_href) = &Psad::buildconf($file);

        ### make sure the configuration is complete
        &check_config();

        %Config = %$Config_href;
        %Cmds   = %$Cmds_href;
        $$mtime_ref = $mtime_tmp;
    }
    return;
}

sub check_config() {
    my @required_vars = qw(
        FW_DATA DISKMOND_CHECK_INTERVAL
        MAX_DISK_PERCENTAGE DISKMOND_PID_FILE
        PSAD_DIR SCAN_DATA_ARCHIVE_DIR
    );
    &Psad::validate_config($CONFIG_FILE, \@required_vars, $Config_href);
    return;
}
